### PR TITLE
NCC: Vazi, Keen Negotiator and support

### LINKED ISSUE #209: Vazi, Keen Negotiator

### PR DESCRIPTION
**Pull request by Northmoc**
closes #209

### CODE REVIEW

File: forge-gui/res/cardsfolder/upcoming/vazi_keen_negotiator.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Name:Vazi, Keen Negotiator
+ManaCost:2 B R G
+Types:Legendary Creature Human Advisor
+PT:3/3
+K:Haste
+A:AB$ Token | Cost$ T | ValidTgts$ Opponent | TokenAmount$ X | TokenScript$ c_a_treasure_sac | TokenOwner$ Targeted | SpellDescription$ Target opponent creates X Treasure tokens, where X is the number of Treasure tokens you created this turn.
+SVar:X:PlayerCountPropertyYou$TreasureTokensCreatedThisTurn
+T:Mode$ SpellCast | ValidCard$ Card | ValidActivatingPlayer$ Opponent | TriggerZones$ Battlefield | ManaFrom$ Treasure | Execute$ TrigPutCounter | TriggerDescription$ Whenever an opponent casts a spell or activates an ability, if mana from a Treasure was spent to cast or activate it, put a +1/+1 counter on target creature, then draw a card.
```
Comment (Hanmac) [inline forge-gui/res/cardsfolder/upcoming/vazi_keen_negotiator.txt:8]: these two trigger can be done as one, i think we have SpellAbilityCast or something

File: forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java [hunk 2]
```diff
@@ -248,6 +248,22 @@ public final boolean performTest(final Map<AbilityKey, Object> runParams) {
             }
         }
 
+        if (hasParam("ManaFrom")) {
+            boolean found = false;
+            for (Mana m : spellAbility.getPayingMana()) {
+                Card source = m.getSourceCard();
+                if (source != null) {
+                    if (source.getType().hasSubtype(getParam("ManaFrom"))) {
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java:256]: Should problably use Valid check

File: forge-gui/res/cardsfolder/upcoming/vazi_keen_negotiator.txt (hunk 1 above)
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/vazi_keen_negotiator.txt:8]: use single SpellAbilityCast trigger :)

File: forge-game/src/main/java/forge/game/player/Player.java [hunk 3]
```diff
@@ -1529,9 +1530,15 @@ else if (discardMadness) {
     public final int getNumTokensCreatedThisTurn() {
         return numTokenCreatedThisTurn;
     }
+    public final int getNumTreasureTokenCreatedThisTurn() {
+        return numTreasureTokenCreatedThisTurn;
+    }
 
     public final void addTokensCreatedThisTurn(Card token) {
         numTokenCreatedThisTurn++;
+        if (token.getType().hasSubtype("Treasure")) {
```
[the remark this answers is not in the document]
Comment (tool4ever) [reply forge-game/src/main/java/forge/game/player/Player.java:1539]: Imo `Count$ThisTurnEntered` should already provide with LKI

so might also get rid of TokensCreatedThisTurn too

[the remark this answers is not in the document]
Comment (Northmoc) [reply forge-game/src/main/java/forge/game/player/Player.java]: I got rid of it as a playerXCount, but couldn't quite get rid of the info on Player due to Akim

One of you might be able to do that, though

[the remark this answers is not in the document]
Comment (Hanmac) [reply forge-game/src/main/java/forge/game/player/Player.java]: I don't think that's correct

Isn't there a card that changes under which control the token enters?

In that case the you create the token, while it enters under opponents control?

[the remark this answers is not in the document]
Comment (tool4ever) [reply forge-game/src/main/java/forge/game/player/Player.java]: yea _Gather Specimens_ can change that

you need to check for owner, then you really created them in accordance with CR 111.2

[the remark this answers is not in the document]
Comment (Hanmac) [reply forge-game/src/main/java/forge/game/player/Player.java]: I was thinking `Crafty Cutpurse` but this changes who creates the token

File: forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java [hunk 4]
```diff
@@ -248,6 +248,22 @@ public final boolean performTest(final Map<AbilityKey, Object> runParams) {
             }
         }
 
+        if (hasParam("ManaFrom")) {
+            boolean found = false;
+            for (Mana m : spellAbility.getPayingMana()) {
+                Card source = m.getSourceCard();
+                if (source != null) {
+                    if (source.isValid(getParam("ManaFrom"), source.getController(), source, spellAbility)) {
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java:256]: shouldn't that be matchValidParam ? and shouldn't it be using the trigger as ctb ?

File: forge-game/src/main/java/forge/game/player/Player.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Northmoc) [reply forge-game/src/main/java/forge/game/player/Player.java]: Using `ThisTurnEntered` + `YouOwn` does solve the _Gather Specimens_ case

**Review by Agetian — APPROVED** [review]

**Pull request merged**